lib/api/apiUtils/object/objectLockHelpers.js

```diff
@@ -126,101 +130,175 @@ function setObjectLockInformation(headers, md, defaultRetention) {
-            return false;
-        }
-        return true;
+    /**
+     * ObjectLockInfo.isComplianceMode
+     * @returns {bool} - True if retention mode is COMPLIANCE
+     */
+    isComplianceMode() {
+        return this.mode === 'COMPLIANCE';
     }
 
-    // an invalid retention change or unrelated to bypass
-    return false;
-}
+    /**
+     * ObjectLockInfo.isExpired
+     * @returns {bool} - True if the retention policy policy has expired
+     */
+    isExpired() {
+        const now = moment();
+        return this.date === null || now.isSameOrAfter(this.date);
+    }
 
-function validateObjectLockUpdate(objectMD, retentionInfo, bypassGovernance) {
-    const { retentionMode: existingMode, retentionDate: existingDateISO } = objectMD;
-    if (!existingMode) {
-        return null;
+    /**
+     * ObjectLockInfo.isExtended
+     * @param {string} timestamp - Timestamp in ISP-8601 format
+     * @returns {bool} - True if the given timestamp is after the policy expiration date or if no expiration date is set
+     */
+    isExtended(timestamp) {
+        return (this.date === null && timestamp !== undefined)
+            || (timestamp !== undefined && moment(timestamp).isAfter(this.date));
     }
 
-    const existingDate = new Date(existingDateISO);
-    const isExpired = existingDate < Date.now();
-    const isExtended = new Date(retentionInfo.date) > existingDate;
+    /**
+     * ObjectLockInfo.canModifyObject
+     * @param {bool} hasGovernanceBypass - Whether to bypass governance retention policies
+     * @returns {bool} - True if the retention policy allows the objects data to be modified (overwritten/deleted)
+     */
+    canModifyObject(hasGovernanceBypass) {
+        return !this.isLocked() || (this.isGovernanceMode() && !!hasGovernanceBypass);
+    }
 
-    if (existingMode === 'GOVERNANCE' && !isExpired && !bypassGovernance) {
-        if (retentionInfo.mode === 'GOVERNANCE' && isExtended) {
-            return null;
+    /**
+     * ObjectLockInfo.canModifyPolicy
+     * @param {object} policyChanges - Proposed changes to the retention policy
+     * @param {bool} hasGovernanceBypass - Whether to bypass governance retention policies
+     * @returns {bool} - True if the changes are allowed to be applied to the retention policy
+     */
+    canModifyPolicy(policyChanges, hasGovernanceBypass) {
+        // If an object does not have a retention policy or it is expired then all changes are allowed
+        if (!this.isLocked()) {
+            return true;
         }
-        return errors.AccessDenied;
-    }
 
-    if (existingMode === 'COMPLIANCE') {
-        if (retentionInfo.mode === 'GOVERNANCE' && !isExpired) {
-            return errors.AccessDenied;
+        // The only allowed change in compliance mode is extending the retention period
+        if (this.isComplianceMode() && policyChanges.mode === 'COMPLIANCE' && this.isExtended(policyChanges.date)) {
+            return true;
```

Comment:
Minor suggestion for clarity

```suggestion
        if (this.isComplianceMode()) {
            if (policyChanges.mode === 'COMPLIANCE' && this.isExtended(policyChanges.date)) {
            return true;
        } else if (this.isGovernanceMode()) {
            ...
        }
```